src/BitIntegers.jl

```diff
@@ -311,14 +311,40 @@ promote_rule(::Type{Float16}, ::Type{<:XBI}) = Float16
 (|)(x::T, y::T) where {T<:XBI} = or_int(x, y)
 xor(x::T, y::T) where {T<:XBI} = xor_int(x, y)
 
->>( x::UBS, y::UBU) = ashr_int(x, y)
->>( x::UBU, y::UBU) = lshr_int(x, y)
->>>(x::UBI, y::UBU) = lshr_int(x, y)
-<<( x::UBI, y::UBU) = shl_int(x, y)
-
->>( x::UBI, y::Int) = ifelse(0 <= y, x >> unsigned(y),  x << unsigned(-y))
-<<( x::UBI, y::Int) = ifelse(0 <= y, x << unsigned(y),  x >> unsigned(-y))
->>>(x::UBI, y::Int) = ifelse(0 <= y, x >>> unsigned(y), x << unsigned(-y))
+# LLVM bit shifting (lshr_int, ashr_int and shl_int) has performance issues for large integers
+# The following functions allows much faster code (10x speedup for 1024 bit integers)
+
+# Note: this should be a power of 2
+const SHIFT_SPLIT_NBITS = 64
+
+@inline @generated function shift_call(sh_fun::Function, x::I, y::UBU) where {I<:XBI}
+    nbits = 8 * sizeof(I)
+    # Performance issue does not afect integers with less than 256 bits
+    nbits <= 128 && return :(sh_fun(x, y))
+    split = SHIFT_SPLIT_NBITS
+    mask = split - 1
+    quote
+        hi = y >> $(trailing_zeros(split))
+        $([:(hi == $val && return sh_fun(sh_fun(x, $(val * split)), y & $mask))
+            for val in 0:div(nbits, split)]...
```

Comment:
This looks correct, but I think I would feel a bit more comfortable merging this if you added a couple more tests testing edge cases, e.g. when `val` doesn't fall in `0:div(nbits, split)]`, or when it's equal to `div(nbits, split)]`, or with e.g. `Int200`. I know this was not tested before, this is now a more complicated algorithm than what was in `Base` (i.e. using directly intrinsics).

Reply:
I agree. I added many more tests in the last commit, it should cover all cases you mentioned.